xarray/core/variable.py

```diff
@@ -427,7 +427,7 @@ def _parse_dimensions(self, dims):
         if isinstance(dims, basestring):
             dims = (dims,)
         dims = tuple(dims)
-        if len(dims) != self.ndim:
+        if self.shape != (0,) and len(dims) != self.ndim:
```

Comment:
This would allow you to create an empty Variable with the wrong number of dimensions, so I don't think this is the right way to fix this.

Something like `xarray.Variable((), [])` should still be an error, because we require labeling all dimensions on a Variable, regardless of their size.

Reply:
Huh, I was thinking that `[]` would be valid with any number of dims. So for example, `xarray.Variable((), [])` would be equivalent to `xarray.Variable((), np.empty(()))`, or `xarray.Variable(('a', 'b'), [])` would be like `xarray.Variable(('a', 'b'), np.empty((0,0)))`.

This interpretation of `[]` would be consistent with the way `to_dict` currently deals with empty variables. For example, `xr.Dataset({'foo': (['a', 'b'], np.empty((0,0)))}).to_dict()['data_vars']['foo']['data']` is `[]`.

But `Variable` seems to interpret `[]` not as having ambiguous dimensionality, but rather `ndim=1`. `to_dict` could instead write `[[]]` for data of shape `(0,0)`, but then there is nothing sensible to write for data of shape `()` that I can think of. Also, `Variable` gives things like `[[]]` or `[[[]]]` leading sizes of 1, like `(1,0)` or `(1,1,0)`, which brings up another issue with trying to infer shape information from empty arrays.

Without changing `Variable`'s current interpretation of the shape of `[]`, `[[]]`, `[[[]]]`, ..., I guess this will have to be explicitly handled in `from_dict`, either locally or by providing a `shape` override on `Variable`.

Reply:
Variable is just casting its inputs with `np.array()` here. Note that `np.array([]).shape == (0,)`.

The culprit seems to be that NumPy's `.tolist()` method is not invertible for size zero arrays. All trailing dimensions seem to be dropped:
```
>>> np.empty((0,)).tolist()
[]
>>> np.empty((0, 2, 3)).tolist()
[]
>>> np.empty((2, 3, 0)).tolist()
[[[], [], []], [[], [], []]]
```

So it appears that we will indeed need to special our logic for restoring these arrays, to take note of the variable's dimensions and their sizes.

Reply:
Here's an especially weird one:
```
>>> np.empty(()).tolist()
1.309863837007e-312
```
As expected, it also shows up here:
```
>>> xr.Dataset({'a': np.empty(())}).to_dict()
{'coords': {}, 'attrs': {}, 'dims': {}, 'data_vars': {'a': {'data': 4.4e-323, 'dims': (), 'attrs': {}}}}
```
From a cursory look it appears that xarray does try to handle 0d arrays, so I guess this is another thing that technically ought to be handled.